test/EFCore.PG.FunctionalTests/Query/CharacterQueryNpgsqlTest.cs

```diff
@@ -0,0 +1,271 @@
+            }
+        }
+
+        #endregion
+
+        #region Support
+
+        /// <summary>
+        /// Provides resources for unit tests.
+        /// </summary>
+        CharacterQueryNpgsqlFixture Fixture { get; }
+
+        /// <summary>
+        /// Initializes resources for unit tests.
+        /// </summary>
+        /// <param name="fixture">The fixture of resources for testing.</param>
+        public CharacterQueryNpgsqlTest(CharacterQueryNpgsqlFixture fixture)
+        {
+            Fixture = fixture;
+            Fixture.TestSqlLoggerFactory.Clear();
+        }
+
+        #endregion
+
+        #region Fixture
+
+        /// <summary>
+        /// Represents a fixture suitable for testing character data.
+        /// </summary>
+        public class CharacterQueryNpgsqlFixture : IDisposable
+        {
+            /// <summary>
+            /// The <see cref="NpgsqlTestStore"/> used for testing.
+            /// </summary>
+            private readonly NpgsqlTestStore _testStore;
+
+            /// <summary>
+            /// The <see cref="DbContextOptions"/> used for testing.
+            /// </summary>
+            private readonly DbContextOptions _options;
+
+            /// <summary>
+            /// The logger factory used for testing.
+            /// </summary>
+            public TestSqlLoggerFactory TestSqlLoggerFactory { get; }
+
+            /// <summary>
+            /// Initializes a <see cref="CharacterQueryNpgsqlTest.CharacterQueryNpgsqlFixture"/>.
+            /// </summary>
+            // ReSharper disable once UnusedMember.Global
+            public CharacterQueryNpgsqlFixture()
+            {
+                TestSqlLoggerFactory = new TestSqlLoggerFactory();
+
+                _testStore = NpgsqlTestStore.CreateScratch();
+
+                _options =
+                    new DbContextOptionsBuilder()
+                        .UseNpgsql(_testStore.ConnectionString, b => b.ApplyConfiguration())
+                        .UseInternalServiceProvider(
+                            new ServiceCollection()
+                                .AddEntityFrameworkNpgsql()
+                                .AddSingleton<ILoggerFactory>(TestSqlLoggerFactory)
+                                .BuildServiceProvider())
+                        .Options;
+
+                using (var context = CreateContext())
+                {
+                    context.Database.EnsureCreated();
+                }
+            }
+
+            /// <summary>
+            /// Creates a new <see cref="CharacterContext"/>.
+            /// </summary>
+            /// <returns>
+            /// A <see cref="CharacterContext"/> for testing.
+            /// </returns>
+            public CharacterContext CreateContext() => new CharacterContext(_options);
+
+            /// <summary>
+            /// Clears the entities in the context.
+            /// </summary>
+            public void ClearEntities()
```

Comment:
It seems like all the tests clear the entities as their first step... It's usually better to find a data set that simply all tests can work with. You can use values or another column to filter out the exact rows you want to access in a given test. This is the way all other EF Core tests work.

Reply:
This seemed like the simplest way forward, since the tests are trying to tease out edge cases (like in #567). Specifically, we want to test that we see PostgreSQL-like behavior on both sides of the insert (when in the local cache and when not), running each test with (effectively) a clean database seemed most appropriate.

I'll give it another pass and see if this can be done in a more traditional way. (It's possible I'm overthinking the isolation requirements.)

Do you know of any other tests that are focused on the insert/local cache/remote cache testing? (From this repo, EF Core, other providers, etc)

Reply:
I'm sure there are some similar tests, but I don't have time to look into it... If you can't find any and think this is the right way forward, go ahead with this - it's not critical.